jetson/nav/stateMachine.hpp

```diff
@@ -44,6 +45,8 @@ class StateMachine : public std::enable_shared_from_this<StateMachine> {
 
     double getDtSeconds();
 
+    std::unique_ptr<Recovery>& getRecovery();
```

Comment:
Generally you never want to have a reference as an instance variable. This code is also undermining the point of a `unique_ptr` (the point is to have a single owner to a heap allocated variable) It helps to watch a couple videos such as https://www.youtube.com/watch?v=UOB7-B2MfwA to understand smart pointers in general.

As far as the references goes, think about it this way. A reference to a unique ptr has no owning factor. So if that unique ptr goes out of scope, you are screwed, you have a reference to garbage memory. This is why you never want to have a reference as a field unless you 100% know it will live past the lifetime of whatever instance has a reference to it. In fact I would argue you never want to do it as it is bad design and prone to bugs.

Reply:
ok thank that makes sense, I should have looked up how to use smart pointers first. In my case the recovery object does need to be shared/used between the stateMachine and the searchStatemachine so should it be a `shared_ptr`? It really only needs to be owned by the stateMachine so I get why it doesn't need to be owned by the searchStateMachine so is there a better way to achieve this and still have it be used by both?

Reply:
Like you said, the owernship doesn't need to be shared. Since search state machine has a pointer to the regular state machine, you can use the recovery object as needed through that without assuming ownership of the unique ptr (don't perform moves). It is perfectly acceptable to use the raw pointer for non-owning functions and logic that just work with the Recovery.

Reply:
this should also be fixed now as I just pass the underlying pointer rather than the ref of the unique_ptr